src/vos/vos_common.c

```diff
@@ -57,6 +57,42 @@ vos_tls_get(void)
 #endif /* VOS_STANDALONE */
 }
 
+/** Add missing timestamp cache entries.  This should be called
+ *  when execution may have been short circuited by a non-existent
+ *  entity so we can fill in the negative timestamps before doing
+ *  timestamp updates.
+ */
+void
+vos_ts_add_missing(struct vos_ts_set *ts_set, daos_key_t *dkey, int akey_nr,
+		   struct vos_akey_data *ad)
+{
+	daos_key_t	*akey;
+	int		 i, rc, remaining;
+
+	if (!vos_ts_in_tx(ts_set) || dkey == NULL)
+		return;
+
+	if (ts_set->ts_etype == VOS_TS_TYPE_DKEY) {
+		/** Add the negative dkey entry */
+		rc = vos_ts_set_add(ts_set, 0 /* don't care */,
+				    dkey->iov_buf,
+				    (int)dkey->iov_len);
+		D_ASSERT(rc == 0);
+	}
+
+	remaining = (VOS_TS_TYPE_AKEY + akey_nr) - ts_set->ts_init_count;
+
+	/** Add negative akey entries */
+	for (i = akey_nr - remaining; i < akey_nr; i++) {
```

Comment:
In such case, akey 0 will already be in the timestamp cache.    This code only affects when we exit the loop early because something doesn't exist.

Reply:
I guess there is the potential for an existent akey to be missed here.

Say we have a per akey condition on akey 0 that causes us to exit the loop.   If akey 1 exists, this loop will use the negative entry for akey1 rather than the entry associated with akey 1.   So it needs a little more work.  But I'd like to address that in a follow up patch.

Reply:
yeah, it's quite complicated and it may be something I address post milestone with a rework of how VOS does things.   Right now, I think we are trying to do too much with overloaded APIs.   I kind of want to split the tree lookups and creations from the check/ilog phase entirely